In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
from sklearn.preprocessing import OneHotEncoder,MultiLabelBinarizer,LabelEncoder,MinMaxScaler,StandardScaler, RobustScaler
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import ADASYN
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense,Dropout
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.models import load_model
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
import numpy as np

In [2]:
data=pd.read_csv('Lending_Loan_Company_Data_Analysis.csv')
data_num=data.select_dtypes(exclude='object')

# Feature Engineering

## Check for Collinearity B/W the Columns

In [3]:
for i in data_num.columns:
    corr=data_num.corr()
    corr[i].sort_values(ascending=False)
    print(corr)

                           Credit_Policy  Interest_Rate  Monthly_Installment  \
Credit_Policy                   1.000000      -0.294089             0.058770   
Interest_Rate                  -0.294089       1.000000             0.276140   
Monthly_Installment             0.058770       0.276140             1.000000   
Natural_log_Annual_Income       0.034906       0.056383             0.448102   
Debt_Income_Ratio              -0.090901       0.220006             0.050202   
FICO_Score                      0.348319      -0.714821             0.086039   
Days_with_Credit_Line           0.099026      -0.124022             0.183297   
Revolving_Balance              -0.187518       0.092527             0.233625   
Revolving_Utilization          -0.104095       0.464837             0.081356   
Enquiries_Last_6_Months        -0.535511       0.202780            -0.010419   
Deliquency_Last_2_Yrs          -0.076318       0.156079            -0.004368   
Public_Derogatory_Records      -0.054243

### Since, Most of the Columns have the Scores less Than 0.5 between them, There is no need to drop off the columns

# Model Building

## 1. Encode the Labels

In [4]:
data_model=data.copy()

In [5]:
le=LabelEncoder()

In [6]:
data_model

,Credit_Policy,Loan_Purpose,Interest_Rate,Monthly_Installment,Natural_log_Annual_Income,Debt_Income_Ratio,FICO_Score,Days_with_Credit_Line,Revolving_Balance,Revolving_Utilization,Enquiries_Last_6_Months,Deliquency_Last_2_Yrs,Public_Derogatory_Records,Paid_Status
0,1,debt_consolidation,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,Paid
1,1,credit_card,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,Paid
2,1,debt_consolidation,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,Paid
3,1,debt_consolidation,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,Paid
4,1,credit_card,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,Paid
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9573,0,all_other,0.1461,344.76,12.180755,10.39,672,10474.000000,215372,82.1,2,0,0,Unpaid
9574,0,all_other,0.1253,257.70,11.141862,0.21,722,4380.000000,184,1.1,5,0,0,Unpaid
9575,0,debt_consolidation,0.1071,97.81,10.596635,13.09,687,3450.041667,10036,82.9,8,0,0,Unpaid
9576,0,home_improvement,0.1600,351.58,10.819778,19.18,692,1800.000000,0,3.2,5,0,0,Unpaid


In [7]:
data['Loan_Purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

In [8]:
data_model['Loan_Purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

In [9]:
data

,Credit_Policy,Loan_Purpose,Interest_Rate,Monthly_Installment,Natural_log_Annual_Income,Debt_Income_Ratio,FICO_Score,Days_with_Credit_Line,Revolving_Balance,Revolving_Utilization,Enquiries_Last_6_Months,Deliquency_Last_2_Yrs,Public_Derogatory_Records,Paid_Status
0,1,debt_consolidation,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,Paid
1,1,credit_card,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,Paid
2,1,debt_consolidation,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,Paid
3,1,debt_consolidation,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,Paid
4,1,credit_card,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,Paid
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9573,0,all_other,0.1461,344.76,12.180755,10.39,672,10474.000000,215372,82.1,2,0,0,Unpaid
9574,0,all_other,0.1253,257.70,11.141862,0.21,722,4380.000000,184,1.1,5,0,0,Unpaid
9575,0,debt_consolidation,0.1071,97.81,10.596635,13.09,687,3450.041667,10036,82.9,8,0,0,Unpaid
9576,0,home_improvement,0.1600,351.58,10.819778,19.18,692,1800.000000,0,3.2,5,0,0,Unpaid


## Description of Labels

### 1. Loan Purpose
#### * debt_consolidation - 2
#### * credit_card - 1
#### * all_other - 0
#### * home_improvement -4
#### * small_business - 6
#### * major_purchase - 5
#### * educational - 3

### 2. Paid Status
#### * Paid - 0
#### * Unpaid - 1

In [10]:
data_model['Installment_to_Income'] = data_model['Monthly_Installment'] / (data['Natural_log_Annual_Income'] / 12)
data_model['Interest_Installment'] = data_model['Interest_Rate'] * data_model['Monthly_Installment']
data_model['Revolving_to_Income'] = data_model['Revolving_Balance'] / data_model['Natural_log_Annual_Income']
data_model['Credit_Length_Years'] = data_model['Days_with_Credit_Line'] / 365
data_model['Risk_Score'] = data_model['Interest_Rate'] / data_model['FICO_Score']

In [11]:
x=data_model.drop(['Paid_Status','Natural_log_Annual_Income'],axis=1)
y=le.fit_transform(data_model['Paid_Status'])

In [12]:
xtrain,xtemp,ytrain,ytemp=train_test_split(x,y,test_size=0.3,random_state=42)
xeval,xtest,yeval,ytest=train_test_split(xtemp,ytemp,test_size=0.5,random_state=42)

In [13]:

cat_cols = ['Credit_Policy', 'Loan_Purpose', 'Enquiries_Last_6_Months' ,'Deliquency_Last_2_Yrs','Public_Derogatory_Records']

# Outlier numeric
robust_cols = [
    'Interest_Rate',
    'Monthly_Installment',
    'FICO_Score',
    'Days_with_Credit_Line',
    'Revolving_Balance',
    'Installment_to_Income',
    'Interest_Installment'
    
]

# Remaining numeric
standard_cols = [col for col in xtrain.columns 
                 if col not in robust_cols + cat_cols]



In [14]:

preprocessor = ColumnTransformer(
    transformers=[
        ('robust', RobustScaler(), robust_cols),
        ('standard', StandardScaler(), standard_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), cat_cols)   
    ]
)

# Step 1: Fit scaler on original train
X_train_scaled = preprocessor.fit_transform(xtrain)

# Step 2: Apply same transform
X_eval = preprocessor.transform(xeval)
X_test = preprocessor.transform(xtest)

# Step 3: Apply ADASYN AFTER scaling
adasyn = ADASYN(sampling_strategy=0.5, random_state=42)
X_train_res_Ada, y_train_res = adasyn.fit_resample(X_train_scaled, ytrain)

In [15]:
model = Sequential()

# Input Layer + Hidden Layer 1
model.add(Dense(64, activation='relu', input_dim=X_train_res_Ada.shape[1]))
model.add(Dropout(0.3))

# Hidden Layer 2
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.2))

# Hidden Layer 3
model.add(Dense(16, activation='relu'))

# Output Layer
model.add(Dense(1, activation='sigmoid'))  # Binary classification

# Compile
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['Precision', 'Recall']
)

# Early Stopping (VERY IMPORTANT)
early_stop = EarlyStopping(
    monitor='val_loss',
    patience=20,
    restore_best_weights=True
)

# Train
history = model.fit(
    X_train_res_Ada, y_train_res,
    validation_data=(X_eval, yeval),
    
    epochs=200,
    batch_size=32,
    callbacks=[early_stop]
    
)

Epoch 1/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 12s 17ms/step - Precision: 0.3874 - Recall: 0.0545 - loss: 0.6349 - val_Precision: 0.0000e+00 - val_Recall: 0.0000e+00 - val_loss: 0.4963
Epoch 2/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 11ms/step - Precision: 0.5616 - Recall: 0.0807 - loss: 0.6129 - val_Precision: 0.3483 - val_Recall: 0.1319 - val_loss: 0.4922
Epoch 3/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 4s 13ms/step - Precision: 0.5415 - Recall: 0.1511 - loss: 0.6052 - val_Precision: 0.3261 - val_Recall: 0.1277 - val_loss: 0.5035
Epoch 4/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 4s 13ms/step - Precision: 0.5763 - Recall: 0.1916 - loss: 0.6002 - val_Precision: 0.3546 - val_Recall: 0.2128 - val_loss: 0.4970
Epoch 5/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 11ms/step - Precision: 0.5761 - Recall: 0.2100 - loss: 0.5967 - val_Precision: 0.3292 - val_Recall: 0.2255 - val_loss: 0.5102
Epoch 6/200
268/268 ━━━━━━━━━━━━━━━━━━━━ 3s 11ms/step - Precision: 0.6057 - Recall: 0.2369 - loss: 0.5928 - val_Precision: 0.3140 - val_Recal

In [16]:
loss, precision, recall = model.evaluate(X_eval, yeval)
print("Loss:", loss)
print("Precision:", precision)
print("Recall:", recall)

45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step - Precision: 0.3551 - Recall: 0.2085 - loss: 0.4840
Loss: 0.4839726388454437
Precision: 0.3550724685192108
Recall: 0.2085106372833252


In [19]:
ypred=model.predict(X_eval)

ypred_t = (model.predict(X_eval) > 0.3).astype(int)

print(classification_report(yeval, ypred_t))


45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 12ms/step
45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step
              precision    recall  f1-score   support

           0       0.88      0.62      0.72      1202
           1       0.22      0.56      0.32       235

    accuracy                           0.61      1437
   macro avg       0.55      0.59      0.52      1437
weighted avg       0.77      0.61      0.66      1437



In [20]:
ypred_Test=model.predict(X_test)

ypred_test = (model.predict(X_test) > 0.3).astype(int)

print(classification_report(ytest, ypred_test))

45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step
45/45 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step
              precision    recall  f1-score   support

           0       0.89      0.62      0.73      1206
           1       0.23      0.60      0.33       231

    accuracy                           0.62      1437
   macro avg       0.56      0.61      0.53      1437
weighted avg       0.78      0.62      0.67      1437



# Insights

The ADASYN-based model significantly improved recall for the 'Unpaid' class, achieving approximately 60% recall on the test dataset. This indicates that the model was able to identify a majority of risky borrowers successfully.

However, the model produced relatively low precision (~23%), meaning that many safe borrowers were incorrectly classified as defaulters. This occurred because:

ADASYN generated synthetic minority-class samples
The classification threshold was lowered to 0.3
The model became more aggressive in predicting the minority class

As a result, the number of false positives increased substantially.

# Business Interpretation

From a business perspective, the ADASYN model is effective for identifying potentially risky loan applicants because it minimizes missed defaulters.

However, the low precision indicates that many low-risk customers may also be flagged as risky, which could:

reduce loan approval rates
affect customer experience
lead to unnecessary rejection of safe applicants

Therefore, although ADASYN improved recall, it created an imbalance between precision and recall, making the predictions less stable for real-world lending applications.

# Summary

The ADASYN approach successfully improved detection of defaulters but introduced excessive false positives due to overprediction of the minority class.

Compared to ADASYN, the class-weight-based ANN model provided a more balanced and stable performance while still maintaining strong detection capability for risky borrowers. Hence, class weights were selected as the preferred imbalance-handling strategy for the final Lending Club loan default prediction model.